src/JsonRpc/InputHandler.cs

```diff
@@ -39,113 +58,234 @@ public class InputHandler : IInputHandler
+                }
+                else
+                {
+                    buffer = buffer.Slice(buffer.GetPosition(1, buffer.PositionOf((byte) '\n') ?? buffer.End));
+                }
+            } while (true);
+        }
+
+        internal async Task ProcessInputStream(CancellationToken cancellationToken)
+        {
+            // some time to attach a debugger
+            // System.Threading.Thread.Sleep(TimeSpan.FromSeconds(5));
+
+            var headersParsed = false;
+            long length = 0;
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var result = await _pipeReader.ReadAsync(cancellationToken);
+                var buffer = result.Buffer;
+
+                if (!headersParsed)
+                {
+                    if (TryParseHeaders(ref buffer, out var line))
+                    {
+                        if (TryParseContentLength(ref line, out length))
                         {
-                            var n = _input.Read(requestBuffer, received, requestBuffer.Length - received);
-                            if (n == 0) return; // no more _input
-                            received += n;
+                            headersParsed = true;
                         }
-                        // TODO sometimes: encoding should be based on the respective header (including the wrong "utf8" value)
-                        var payload = System.Text.Encoding.UTF8.GetString(requestBuffer);
-                        HandleRequest(payload, CancellationToken.None);
                     }
                 }
-                catch (IOException)
+
+                if (headersParsed && length == 0)
+                {
+                    HandleRequest(new ReadOnlySequence<byte>(Array.Empty<byte>()));
+                    headersParsed = false;
+                }
+
+                if (headersParsed)
+                {
+                    if (TryParseBodyString(length, ref buffer, out var line))
+                    {
+                        headersParsed = false;
+                        length = 0;
+                        HandleRequest(line);
+                    }
+                }
+
+                _pipeReader.AdvanceTo(buffer.Start, buffer.End);
+
+                // Stop reading if there's no more data coming.
+                if (result.IsCompleted && buffer.GetPosition(0, buffer.Start).Equals(buffer.GetPosition(0, buffer.End)))
                 {
-                    _logger.LogError("Input stream has been closed.");
                     break;
                 }
             }
         }
 
-        private void HandleRequest(string request, CancellationToken cancellationToken)
+        private void HandleRequest(in ReadOnlySequence<byte> request)
         {
             JToken payload;
             try
             {
-                payload = JToken.Parse(request);
+                using var textReader = new StreamReader(request.AsStream());
```

Comment:
If you could get rid of this you can make reduce allocations even further.

Reply:
@davidfowl baby steps my friend!

Pipelines first, then I have this branch where I explored moving to system.text.json and it mostly works.  I have to however refactor the language client as it doesn't really mesh well with the server at this point.
https://github.com/OmniSharp/csharp-language-server-protocol/tree/feature/systemtextjson

Reply:
And by works I mean it compiles 😆 (without the client project)